File: minigit-5/ts/src/commands/merge.ts
// merge command - Merge branch into current branch

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, ensureDir, formatAuthorDate, getAuthorInfo, getCommitterInfo } from '../utils';
import { readIndex, writeIndex } from '../index-file';
import { getHeadCommit, getCurrentBranch, resolveRef, resolveRevision, updateRef, setHead, isDetachedHead } from '../refs';
import { walkTree, getTreeFromTreeIsh, getBlob, createCommit, createTree, createBlob, getCommit } from '../objects';
import { findMergeBase, mergeFiles } from '../merge-algo';
import { IndexEntry } from '../types';

export function merge(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  let noCommit = false;
  let abortMerge = false;
  let targetBranch: string | null = null;

  for (const arg of args) {
    if (arg === '--no-commit') {
      noCommit = true;
    } else if (arg === '--abort') {
      abortMerge = true;
    } else if (!arg.startsWith('-')) {
      targetBranch = arg;
    }
  }

  // Handle abort
  if (abortMerge) {
    const mergeHeadPath = path.join(repoRoot, '.minigit', 'MERGE_HEAD');
    if (fs.existsSync(mergeHeadPath)) {
      // Reset to HEAD
      const headSha = getHeadCommit(repoRoot);
      if (headSha) {
        // Reset working tree and index to HEAD
        const treeSha = getTreeFromTreeIsh(headSha, repoRoot);
        resetToTree(treeSha, repoRoot);
      }
      fs.unlinkSync(mergeHeadPath);
      const mergeMsgPath = path.join(repoRoot, '.minigit', 'MERGE_MSG');
      if (fs.existsSync(mergeMsgPath)) {
        fs.unlinkSync(mergeMsgPath);
      }
      console.log('Merge aborted.');
      return 0;
    } else {
      console.error('error: There is no merge in progress');
      return 1;
    }
  }

  if (!targetBranch) {
    console.error('fatal: no branch specified');
    return 1;
  }

  // Resolve target
  let targetSha: string;
  try {
    // Try as branch first
    const branchSha = resolveRef(`refs/heads/${targetBranch}`, repoRoot);
    if (branchSha) {
      targetSha = branchSha;
    } else {
      targetSha = resolveRevision(targetBranch, repoRoot);
    }
  } catch (e: any) {
    console.error(`merge: ${targetBranch} - not something we can merge`);
    return 1;
  }

  // Get current HEAD
  const headSha = getHeadCommit(repoRoot);
  if (!headSha) {
    console.error('fatal: HEAD does not point to a valid commit');
    return 1;
  }

  // Check if already up to date
  if (headSha === targetSha) {
    console.log('Already up to date.');
    return 0;
  }

  // Check if target is ancestor of HEAD (already merged)
  if (isAncestor(targetSha, headSha, repoRoot)) {
    console.log('Already up to date.');
    return 0;
  }

  // Check if HEAD is ancestor of target (fast-forward possible)
  if (isAncestor(headSha, targetSha, repoRoot)) {
    // Fast-forward
    const result = updateWorkingTree(targetSha, repoRoot);
    if (result !== 0) return result;

    if (!noCommit) {
      const branch = getCurrentBranch(repoRoot);
      if (branch) {
        updateRef(`refs/heads/${branch}`, targetSha, repoRoot);
      } else {
        setHead(targetSha, repoRoot);
      }

      console.log(`Updating ${headSha.slice(0, 7)}..${targetSha.slice(0, 7)}`);
      console.log('Fast-forward');
    } else {
      // For --no-commit, save merge state so status shows changes
      const mergeHeadPath = path.join(repoRoot, '.minigit', 'MERGE_HEAD');
      fs.writeFileSync(mergeHeadPath, targetSha + '\n');
    }
    return 0;
  }

  // Three-way merge
  const mergeBase = findMergeBase(headSha, targetSha, repoRoot);
  if (!mergeBase) {
    console.error('fatal: refusing to merge unrelated histories');
    return 1;
  }

  // Get trees
  const baseTree = getTreeFromTreeIsh(mergeBase, repoRoot);
  const headTree = getTreeFromTreeIsh(headSha, repoRoot);
  const targetTree = getTreeFromTreeIsh(targetSha, repoRoot);

  const baseFiles = walkTree(baseTree, '', repoRoot);
  const headFiles = walkTree(headTree, '', repoRoot);
  const targetFiles = walkTree(targetTree, '', repoRoot);

  // Collect all files
  const allPaths = new Set([...baseFiles.keys(), ...headFiles.keys(), ...targetFiles.keys()]);

  const conflicts: string[] = [];
  const mergedEntries: Array<{ path: string; sha: string; mode: string }> = [];

  for (const filePath of allPaths) {
    const baseEntry = baseFiles.get(filePath);
    const headEntry = headFiles.get(filePath);
    const targetEntry = targetFiles.get(filePath);

    // Skip if unchanged
    if (headEntry && targetEntry && headEntry.sha === targetEntry.sha) {
      mergedEntries.push({ path: filePath, sha: headEntry.sha, mode: headEntry.mode });
      continue;
    }

    // File only in HEAD (not in target)
    if (headEntry && !targetEntry) {
      if (!baseEntry) {
        // New in HEAD, keep it
        mergedEntries.push({ path: filePath, sha: headEntry.sha, mode: headEntry.mode });
      } else if (baseEntry.sha === headEntry.sha) {
        // Deleted in target, unchanged in HEAD - delete
        // Don't add to merged
      } else {
        // Modified in HEAD, deleted in target - conflict
        conflicts.push(filePath);
        mergedEntries.push({ path: filePath, sha: headEntry.sha, mode: headEntry.mode });
      }
      continue;
    }

    // File only in target (not in HEAD)
    if (targetEntry && !headEntry) {
      if (!baseEntry) {
        // New in target, add it
        mergedEntries.push({ path: filePath, sha: targetEntry.sha, mode: targetEntry.mode });
      } else if (baseEntry.sha === targetEntry.sha) {
        // Deleted in HEAD, unchanged in target - delete
        // Don't add to merged
      } else {
        // Modified in target, deleted in HEAD - conflict
        conflicts.push(filePath);
        mergedEntries.push({ path: filePath, sha: targetEntry.sha, mode: targetEntry.mode });
      }
      continue;
    }

    // Both have the file
    if (headEntry && targetEntry) {
      if (!baseEntry) {
        // New in both - need to merge
        const headContent = getBlob(headEntry.sha, repoRoot).toString();
        const targetContent = getBlob(targetEntry.sha, repoRoot).toString();

        if (headContent === targetContent) {
          mergedEntries.push({ path: filePath, sha: headEntry.sha, mode: headEntry.mode });
        } else {
          // Conflict
          const merged = mergeFiles(null, headContent, targetContent, 'HEAD', targetBranch);
          const sha = createBlob(Buffer.from(merged.mergedContent || ''), true, repoRoot);
          mergedEntries.push({ path: filePath, sha, mode: headEntry.mode });
          if (!merged.success) {
            conflicts.push(filePath);
          }
        }
        continue;
      }

      // Modified in one or both
      const baseContent = getBlob(baseEntry.sha, repoRoot).toString();
      const headContent = getBlob(headEntry.sha, repoRoot).toString();
      const targetContent = getBlob(targetEntry.sha, repoRoot).toString();

      if (baseEntry.sha === headEntry.sha) {
        // Only modified in target
        mergedEntries.push({ path: filePath, sha: targetEntry.sha, mode: targetEntry.mode });
      } else if (baseEntry.sha === targetEntry.sha) {
        // Only modified in HEAD
        mergedEntries.push({ path: filePath, sha: headEntry.sha, mode: headEntry.mode });
      } else {
        // Modified in both - need three-way merge
        const merged = mergeFiles(baseContent, headContent, targetContent, 'HEAD', targetBranch);
        const sha = createBlob(Buffer.from(merged.mergedContent || ''), true, repoRoot);
        mergedEntries.push({ path: filePath, sha, mode: headEntry.mode });
        if (!merged.success) {
          conflicts.push(filePath);
        }
      }
    }
  }

  // Write merged files to working tree and index
  const index = readIndex(repoRoot);
  index.entries = [];

  for (const entry of mergedEntries) {
    const fullPath = path.join(repoRoot, entry.path);
    ensureDir(path.dirname(fullPath));

    const content = getBlob(entry.sha, repoRoot);
    fs.writeFileSync(fullPath, content);

    if (entry.mode === '100755') {
      fs.chmodSync(fullPath, 0o755);
    }

    const stats = fs.statSync(fullPath);
    const indexEntry: IndexEntry = {
      ctimeSec: Math.floor(stats.ctimeMs / 1000),
      ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stats.mtimeMs / 1000),
      mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
      dev: stats.dev,
      ino: stats.ino,
      mode: parseInt(entry.mode, 8),
      uid: stats.uid,
      gid: stats.gid,
      size: stats.size,
      sha: entry.sha,
      flags: Math.min(entry.path.length, 0xfff),
      path: entry.path,
    };
    index.entries.push(indexEntry);
  }

  writeIndex(index, repoRoot);

  // Delete files not in merged set
  const headFilesSet = new Set(headFiles.keys());
  for (const filePath of headFilesSet) {
    if (!mergedEntries.find(e => e.path === filePath)) {
      const fullPath = path.join(repoRoot, filePath);
      if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
      }
    }
  }

  // Handle conflicts
  if (conflicts.length > 0) {
    // Save merge state
    const mergeHeadPath = path.join(repoRoot, '.minigit', 'MERGE_HEAD');
    fs.writeFileSync(mergeHeadPath, targetSha + '\n');

    const mergeMsgPath = path.join(repoRoot, '.minigit', 'MERGE_MSG');
    fs.writeFileSync(mergeMsgPath, `Merge branch '${targetBranch}'\n\nConflicts:\n${conflicts.map(c => `\t${c}`).join('\n')}\n`);

    console.error('Auto-merging failed; fix conflicts and then commit the result.');
    console.error('Automatic merge failed; fix conflicts and then commit the result.');
    return 1;
  }

  // Create merge commit (unless --no-commit)
  if (!noCommit) {
    // Build tree from index
    const treeSha = buildTreeFromEntries(mergedEntries, repoRoot);

    // Create merge commit
    const authorInfo = getAuthorInfo();
    const committerInfo = getCommitterInfo();
    const author = formatAuthorDate(authorInfo.name, authorInfo.email, authorInfo.date, authorInfo.tz);
    const committer = formatAuthorDate(committerInfo.name, committerInfo.email, committerInfo.date, committerInfo.tz);

    const message = `Merge branch '${targetBranch}'`;
    const commitSha = createCommit(treeSha, [headSha, targetSha], author, committer, message, repoRoot);

    // Update ref
    const branch = getCurrentBranch(repoRoot);
    if (branch) {
      updateRef(`refs/heads/${branch}`, commitSha, repoRoot);
    } else {
      setHead(commitSha, repoRoot);
    }

    console.log(`Merge made by the 'ort' strategy.`);
  }

  return 0;
}

function isAncestor(ancestor: string, descendant: string, repoRoot: string): boolean {
  const visited = new Set<string>();
  const queue = [descendant];

  while (queue.length > 0) {
    const sha = queue.shift()!;
    if (visited.has(sha)) continue;
    visited.add(sha);

    if (sha === ancestor) return true;

    try {
      const commit = getCommit(sha, repoRoot);
      queue.push(...commit.parents);
    } catch {
      // Invalid
    }
  }

  return false;
}

function updateWorkingTree(targetSha: string, repoRoot: string): number {
  const targetTree = getTreeFromTreeIsh(targetSha, repoRoot);
  resetToTree(targetTree, repoRoot);
  return 0;
}

function resetToTree(treeSha: string, repoRoot: string): void {
  const files = walkTree(treeSha, '', repoRoot);
  const index = readIndex(repoRoot);

  // Remove files not in tree
  const currentFiles = new Set<string>();
  collectCurrentFiles(repoRoot, repoRoot, currentFiles);

  for (const filePath of currentFiles) {
    if (!files.has(filePath)) {
      const fullPath = path.join(repoRoot, filePath);
      if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
      }
    }
  }

  // Write files from tree
  index.entries = [];

  for (const [filePath, entry] of files) {
    const fullPath = path.join(repoRoot, filePath);
    ensureDir(path.dirname(fullPath));

    const content = getBlob(entry.sha, repoRoot);
    fs.writeFileSync(fullPath, content);

    if (entry.mode === '100755') {
      fs.chmodSync(fullPath, 0o755);
    }

    const stats = fs.statSync(fullPath);
    const indexEntry: IndexEntry = {
      ctimeSec: Math.floor(stats.ctimeMs / 1000),
      ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stats.mtimeMs / 1000),
      mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
      dev: stats.dev,
      ino: stats.ino,
      mode: parseInt(entry.mode, 8),
      uid: stats.uid,
      gid: stats.gid,
      size: stats.size,
      sha: entry.sha,
      flags: Math.min(filePath.length, 0xfff),
      path: filePath,
    };
    index.entries.push(indexEntry);
  }

  writeIndex(index, repoRoot);
}

function collectCurrentFiles(dir: string, repoRoot: string, result: Set<string>): void {
  try {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name === '.minigit') continue;
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        collectCurrentFiles(fullPath, repoRoot, result);
      } else {
        const relativePath = path.relative(repoRoot, fullPath).split(path.sep).join('/');
        result.add(relativePath);
      }
    }
  } catch {
    // Ignore errors
  }
}

function buildTreeFromEntries(entries: Array<{ path: string; sha: string; mode: string }>, repoRoot: string): string {
  // Group by directory
  const root: Map<string, { sha: string; mode: string } | Map<string, any>> = new Map();

  for (const entry of entries) {
    const parts = entry.path.split('/');
    let current = root;

    for (let i = 0; i < parts.length - 1; i++) {
      if (!current.has(parts[i])) {
        current.set(parts[i], new Map());
      }
      current = current.get(parts[i]) as Map<string, any>;
    }

    current.set(parts[parts.length - 1], { sha: entry.sha, mode: entry.mode });
  }

  return buildTreeRecursive(root, repoRoot);
}

function buildTreeRecursive(dir: Map<string, any>, repoRoot: string): string {
  const treeEntries: Array<{ mode: string; name: string; sha: string }> = [];

  for (const [name, value] of dir) {
    if (value instanceof Map) {
      // Directory
      const subTreeSha = buildTreeRecursive(value, repoRoot);
      treeEntries.push({ mode: '40000', name, sha: subTreeSha });
    } else {
      // File
      treeEntries.push({ mode: value.mode, name, sha: value.sha });
    }
  }

  return createTree(treeEntries, repoRoot);
}

File: minigit-5/py/commands/merge.py
"""merge command - Merge branches."""

import sys
import os
from pathlib import Path
from datetime import datetime, timezone

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_file_mode
from objects import Blob, Tree, Commit
from index import Index, create_entry_from_file
from refs import (
    resolve_head, resolve_revision, get_current_branch, get_head_ref,
    write_ref, read_ref
)
from merge_algo import three_way_merge, find_merge_base


def run(args: list[str]) -> int:
    """Merge branches."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    no_commit = False
    abort = False
    branch = None

    for arg in args:
        if arg == '--no-commit':
            no_commit = True
        elif arg == '--abort':
            abort = True
        elif not arg.startswith('-'):
            branch = arg

    if abort:
        return abort_merge(repo_root)

    if not branch:
        print("fatal: branch name required", file=sys.stderr)
        return 1

    # Resolve branch to SHA
    target_sha = resolve_revision(repo_root, branch)
    if target_sha is None:
        print(f"fatal: '{branch}' is not a valid branch name", file=sys.stderr)
        return 1

    head_sha = resolve_head(repo_root)
    if head_sha is None:
        print("fatal: cannot merge into empty repository", file=sys.stderr)
        return 1

    # Check if already up to date
    if head_sha == target_sha:
        print("Already up to date.")
        return 0

    # Check if target is ancestor of HEAD
    if is_ancestor(repo_root, target_sha, head_sha):
        print("Already up to date.")
        return 0

    # Check if HEAD is ancestor of target (fast-forward)
    if is_ancestor(repo_root, head_sha, target_sha):
        return fast_forward(repo_root, target_sha, branch, no_commit)

    # Need to do a real merge
    return three_way_merge_branches(repo_root, head_sha, target_sha, branch, no_commit)


def fast_forward(repo_root: Path, target_sha: str, branch: str, no_commit: bool) -> int:
    """Perform a fast-forward merge."""
    from commands.checkout import update_working_tree

    update_working_tree(repo_root, target_sha)

    if not no_commit:
        head_ref = get_head_ref(repo_root)
        if head_ref:
            write_ref(repo_root, head_ref, target_sha)

    print(f"Fast-forward to {target_sha[:7]}")
    return 0


def three_way_merge_branches(repo_root: Path, head_sha: str, target_sha: str, branch: str, no_commit: bool) -> int:
    """Perform a three-way merge."""
    # Find merge base
    base_sha = find_merge_base(repo_root, head_sha, target_sha)
    if base_sha is None:
        print("fatal: refusing to merge unrelated histories", file=sys.stderr)
        return 1

    # Get file contents from each version
    base_files = get_tree_files(repo_root, Commit.read(base_sha, repo_root).tree_sha)
    head_files = get_tree_files(repo_root, Commit.read(head_sha, repo_root).tree_sha)
    target_files = get_tree_files(repo_root, Commit.read(target_sha, repo_root).tree_sha)

    # Merge each file
    all_files = set(base_files.keys()) | set(head_files.keys()) | set(target_files.keys())
    index = Index.read(repo_root)
    has_conflicts = False

    for name in sorted(all_files):
        base_sha_file = base_files.get(name, (None, None))[0]
        head_sha_file = head_files.get(name, (None, None))[0]
        target_sha_file = target_files.get(name, (None, None))[0]

        head_mode = head_files.get(name, (None, 0o100644))[1]
        target_mode = target_files.get(name, (None, 0o100644))[1]
        mode = head_mode or target_mode or 0o100644

        conflict = merge_file(
            repo_root, name,
            base_sha_file, head_sha_file, target_sha_file,
            index, mode, branch
        )
        if conflict:
            has_conflicts = True

    index.write(repo_root)

    if has_conflicts:
        # Write merge state
        merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
        merge_head.write_text(target_sha + '\n')
        print("Automatic merge failed; fix conflicts and then commit the result.")
        return 1

    if no_commit:
        print("Automatic merge went well; stopped before committing as requested")
        return 0

    # Create merge commit
    return create_merge_commit(repo_root, head_sha, target_sha, branch, index)


def merge_file(repo_root: Path, name: str,
               base_sha: str, head_sha: str, target_sha: str,
               index: Index, mode: int, branch: str) -> bool:
    """Merge a single file. Returns True if there's a conflict."""
    file_path = repo_root / name

    # Handle cases where file doesn't exist in one or more versions
    if head_sha is None and target_sha is None:
        # File deleted in both - nothing to do
        if file_path.exists():
            file_path.unlink()
        index.remove_entry(name)
        return False

    if head_sha is None:
        # File added or exists only in target
        blob = Blob.read(target_sha, repo_root)
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_bytes(blob.data)
        entry = create_entry_from_file(file_path, name, target_sha, mode)
        index.add_entry(entry)
        return False

    if target_sha is None:
        # File exists only in head - keep it (unless it was deleted in target)
        if base_sha and base_sha == head_sha:
            # File was deleted in target, unchanged in head - delete it
            if file_path.exists():
                file_path.unlink()
            index.remove_entry(name)
        # Otherwise keep head version
        return False

    if head_sha == target_sha:
        # Same content - no merge needed
        return False

    if base_sha == head_sha:
        # Head unchanged, take target
        blob = Blob.read(target_sha, repo_root)
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_bytes(blob.data)
        entry = create_entry_from_file(file_path, name, target_sha, mode)
        index.add_entry(entry)
        return False

    if base_sha == target_sha:
        # Target unchanged, keep head
        return False

    # Need to do content merge
    base_content = Blob.read(base_sha, repo_root).data if base_sha else b''
    head_content = Blob.read(head_sha, repo_root).data
    target_content = Blob.read(target_sha, repo_root).data

    # Check for binary
    if is_binary(head_content) or is_binary(target_content):
        # Binary conflict - keep head, report conflict
        print(f"warning: Cannot merge binary file {name}")
        return True

    # Text merge
    base_lines = base_content.decode('utf-8', errors='replace').split('\n')
    head_lines = head_content.decode('utf-8', errors='replace').split('\n')
    target_lines = target_content.decode('utf-8', errors='replace').split('\n')

    # Remove trailing empty line from split
    if base_lines and base_lines[-1] == '':
        base_lines = base_lines[:-1]
    if head_lines and head_lines[-1] == '':
        head_lines = head_lines[:-1]
    if target_lines and target_lines[-1] == '':
        target_lines = target_lines[:-1]

    result = three_way_merge(base_lines, head_lines, target_lines, "HEAD", branch)

    file_path.parent.mkdir(parents=True, exist_ok=True)
    file_path.write_text(result.content)

    if not result.has_conflicts:
        blob = Blob(result.content.encode())
        blob.write(repo_root)
        entry = create_entry_from_file(file_path, name, blob.sha, mode)
        index.add_entry(entry)

    return result.has_conflicts


def create_merge_commit(repo_root: Path, head_sha: str, target_sha: str, branch: str, index: Index) -> int:
    """Create a merge commit."""
    from commands.commit import build_tree, format_timestamp

    tree_sha = build_tree(repo_root, index)

    author = os.environ.get('GIT_AUTHOR_NAME', 'Unknown')
    author_email = os.environ.get('GIT_AUTHOR_EMAIL', 'unknown@example.com')
    committer = os.environ.get('GIT_COMMITTER_NAME', author)
    committer_email = os.environ.get('GIT_COMMITTER_EMAIL', author_email)
    author_date = format_timestamp(os.environ.get('GIT_AUTHOR_DATE'))
    committer_date = format_timestamp(os.environ.get('GIT_COMMITTER_DATE'))

    message = f"Merge branch '{branch}'"

    commit = Commit(
        tree_sha=tree_sha,
        parents=[head_sha, target_sha],
        author=author,
        author_email=author_email,
        author_date=author_date,
        committer=committer,
        committer_email=committer_email,
        committer_date=committer_date,
        message=message
    )
    commit_sha = commit.write(repo_root)

    head_ref = get_head_ref(repo_root)
    if head_ref:
        write_ref(repo_root, head_ref, commit_sha)

    # Clean up merge state
    merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
    if merge_head.exists():
        merge_head.unlink()

    current = get_current_branch(repo_root)
    print(f"[{current} {commit_sha[:7]}] {message}")
    return 0


def abort_merge(repo_root: Path) -> int:
    """Abort an in-progress merge."""
    merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
    if not merge_head.exists():
        print("fatal: there is no merge to abort", file=sys.stderr)
        return 1

    # Reset to HEAD
    head_sha = resolve_head(repo_root)
    if head_sha:
        from commands.checkout import update_working_tree
        update_working_tree(repo_root, head_sha)

    merge_head.unlink()
    print("Merge aborted.")
    return 0


def get_tree_files(repo_root: Path, tree_sha: str, prefix: str = '') -> dict:
    """Recursively get all files from a tree."""
    files = {}
    tree = Tree.read(tree_sha, repo_root)

    for entry in tree.entries:
        full_path = f"{prefix}{entry.name}" if prefix else entry.name
        if entry.mode == 0o40000:
            files.update(get_tree_files(repo_root, entry.sha, full_path + '/'))
        else:
            files[full_path] = (entry.sha, entry.mode)

    return files


def is_ancestor(repo_root: Path, commit1: str, commit2: str) -> bool:
    """Check if commit1 is an ancestor of commit2."""
    visited = set()
    queue = [commit2]

    while queue:
        sha = queue.pop(0)
        if sha in visited:
            continue
        visited.add(sha)

        if sha == commit1:
            return True

        try:
            commit = Commit.read(sha, repo_root)
            queue.extend(commit.parents)
        except:
            pass

    return False


def is_binary(data: bytes) -> bool:
    """Check if data appears to be binary."""
    sample = data[:8000]
    return b'\x00' in sample
